test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -79,167 +87,320 @@ func (s *snapshottableTestSuite) SkipRedundantSuite(driver TestDriver, pattern t
-			framework.ExpectNoError(dc.Resource(SnapshotClassGVR).Delete(context.TODO(), vsc.GetName(), metav1.DeleteOptions{}))
-		}()
+	ginkgo.BeforeEach(func() {
+		// Check preconditions.
+		framework.ExpectEqual(pattern.SnapshotType, testpatterns.DynamicCreatedSnapshot)
+		dInfo := driver.GetDriverInfo()
+		ok := false
+		sDriver, ok = driver.(SnapshottableTestDriver)
+		if !dInfo.Capabilities[CapSnapshotDataSource] || !ok {
+			e2eskipper.Skipf("Driver %q does not support snapshots - skipping", dInfo.Name)
+		}
+		dDriver, ok = driver.(DynamicPVTestDriver)
+		if !ok {
+			e2eskipper.Skipf("Driver %q does not support dynamic provisioning - skipping", driver.GetDriverInfo().Name)
+		}
+	})
 
-		ginkgo.By("creating a snapshot")
-		snapshot := getSnapshot(pvc.Name, pvc.Namespace, vsc.GetName())
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	ginkgo.It("should create snapshot with delete policy [Feature:VolumeSnapshotDataSource]", func() {
+		l := &snapshottableLocal{}
 
-		snapshot, err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Create(context.TODO(), snapshot, metav1.CreateOptions{})
-		framework.ExpectNoError(err)
-		defer func() {
-			framework.Logf("deleting snapshot %q/%q", snapshot.GetNamespace(), snapshot.GetName())
-			// typically this snapshot has already been deleted
-			err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Delete(context.TODO(), snapshot.GetName(), metav1.DeleteOptions{})
-			if err != nil && !apierrors.IsNotFound(err) {
-				framework.Failf("Error deleting snapshot %q. Error: %v", pvc.Name, err)
-			}
-		}()
-		err = WaitForSnapshotReady(dc, snapshot.GetNamespace(), snapshot.GetName(), framework.Poll, framework.SnapshotCreateTimeout)
-		framework.ExpectNoError(err)
+		init(l)
+		defer cleanup(l)
 
-		ginkgo.By("checking the snapshot")
-		// Get new copy of the snapshot
-		snapshot, err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Get(context.TODO(), snapshot.GetName(), metav1.GetOptions{})
-		framework.ExpectNoError(err)
+		sct := SnapshotClassTest{
+			DeletionPolicy: "Delete",
+		}
 
-		// Get the bound snapshotContent
-		snapshotStatus := snapshot.Object["status"].(map[string]interface{})
-		snapshotContentName := snapshotStatus["boundVolumeSnapshotContentName"].(string)
-		snapshotContent, err := dc.Resource(SnapshotContentGVR).Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
-		framework.ExpectNoError(err)
+		TestSnapshottable(l, sct)
+		TestSnapshotDeleted(l, sct)
+
+	})
+
+	ginkgo.It("should not delete snapshot with retain policy [Feature:VolumeSnapshotDataSource]", func() {
+		l := &snapshottableLocal{}
+
+		init(l)
+		defer cleanup(l)
+
+		sct := SnapshotClassTest{
+			DeletionPolicy: "Retain",
+		}
```

Comment:
need not be defined as a separate struct. We can directly define a struct SnapshotClassTest instance  in the function call.